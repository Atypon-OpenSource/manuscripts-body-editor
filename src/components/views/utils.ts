/*!
 * Â© 2022 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  Designation,
  designationWithFileSectionsMap,
  extensionsWithFileTypesMap,
  FileSectionType,
  FileType,
  fileTypesWithIconMap,
  namesWithDesignationMap,
} from '@manuscripts/style-guide'

import { SubmissionAttachment } from '../../views/FigureComponent'
import { ExternalFileIcon } from './FilesDropdown'

export const getIcon = (file: SubmissionAttachment) => {
  const fileExtension = file.name.split('.').pop() || ''
  const fileType = extensionsWithFileTypesMap.get(fileExtension.toLowerCase())
  return fileTypesWithIconMap.get(fileType)
}

const getDesignationSectionType = (
  designation: Designation,
  fileType?: FileType,
  mediaAlternativesEnabled?: boolean
) => {
  if (mediaAlternativesEnabled) {
    // TODO:: specify file type for media Alternatives
    return designationWithFileSectionsMap.get(designation)
  } else {
    return (
      fileType === FileType.Image &&
      designationWithFileSectionsMap.get(designation)
    )
  }
}

export const getPaperClipButtonFiles = (
  externalFiles?: SubmissionAttachment[],
  mediaAlternativesEnabled?: boolean
) => {
  const files: {
    supplements: ExternalFileIcon[]
    otherFiles: ExternalFileIcon[]
  } = {
    supplements: [],
    otherFiles: [],
  }

  externalFiles?.map((file) => {
    const designation = namesWithDesignationMap.get(file.type.label)
    const fileExtension = file.name.split('.').pop() || ''
    const fileType = extensionsWithFileTypesMap.get(fileExtension.toLowerCase())

    if (designation !== undefined) {
      switch (
        getDesignationSectionType(
          designation,
          fileType,
          mediaAlternativesEnabled
        )
      ) {
        case FileSectionType.Supplements:
          files.supplements.push({
            ...file,
            icon: fileTypesWithIconMap.get(fileType),
          })
          break
        case FileSectionType.OtherFile:
          files.otherFiles.push({
            ...file,
            icon: fileTypesWithIconMap.get(fileType),
          })
          break
      }
    }
  })
  return files
}

export const getOtherFiles = (
  externalFiles?: SubmissionAttachment[],
  mediaAlternativesEnabled?: boolean
) =>
  (externalFiles &&
    externalFiles.filter((file) => {
      const designation = namesWithDesignationMap.get(file.type.label)
      const fileExtension = file.name.split('.').pop() || ''
      const fileType = extensionsWithFileTypesMap.get(
        fileExtension.toLowerCase()
      )
      if (designation !== undefined) {
        const sectionType = getDesignationSectionType(
          designation,
          fileType,
          mediaAlternativesEnabled
        )
        return sectionType === FileSectionType.OtherFile
      }
    })) ||
  []
