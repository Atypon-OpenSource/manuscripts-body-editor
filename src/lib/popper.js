/*!
 * Â© 2019 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import Popper from 'popper.js';
export class PopperManager {
    show(target, contents, placement = 'bottom', showArrow = true) {
        if (this.activePopper) {
            return this.destroy();
        }
        window.requestAnimationFrame(() => {
            const container = document.createElement('div');
            container.className = 'popper';
            const modifiers = {};
            if (showArrow) {
                const arrow = document.createElement('div');
                arrow.className = 'popper-arrow';
                container.appendChild(arrow);
                modifiers.arrow = {
                    element: arrow,
                };
            }
            else {
                modifiers.arrow = {
                    enabled: false,
                };
            }
            container.appendChild(contents);
            document.body.appendChild(container);
            this.activePopper = new Popper(target, container, {
                placement,
                removeOnDestroy: true,
                modifiers,
                onCreate: () => {
                    this.focusInput(container);
                },
            });
        });
    }
    destroy() {
        if (this.activePopper) {
            this.activePopper.destroy();
            delete this.activePopper;
        }
    }
    update() {
        if (this.activePopper) {
            this.activePopper.update();
        }
    }
    focusInput(container) {
        const element = container.querySelector('input');
        if (element) {
            element.focus();
        }
    }
}
