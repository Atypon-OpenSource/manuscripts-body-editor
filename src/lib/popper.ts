/*!
 * Â© 2019 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  createPopper,
  Instance,
  Placement,
  StrictModifiers,
} from '@popperjs/core'

export class PopperManager {
  private activePopper?: Instance
  private handleDocumentClick?: (e: Event) => void

  public show(
    target: Element,
    contents: HTMLElement,
    placement: Placement = 'bottom',
    showArrow = true,
    modifiers: Array<Partial<StrictModifiers>> = []
  ) {
    // destroy any existing popper first
    // checking activePopper is in destroy() method
    this.destroy()

    window.requestAnimationFrame(() => {
      const container = document.createElement('div')
      container.className = 'popper'

      container.addEventListener('click', (e) => {
        e.stopPropagation()
      })

      if (showArrow) {
        const arrow = document.createElement('div')
        arrow.className = 'popper-arrow'
        container.appendChild(arrow)
        modifiers.push({
          name: 'arrow',
          options: {
            element: arrow,
          },
        })
      } else {
        modifiers.push({
          name: 'arrow',
          options: {
            element: null,
          },
        })
      }

      container.appendChild(contents)
      document.body.appendChild(container)

      this.activePopper = createPopper(target, container, {
        placement,
        modifiers,
        onFirstUpdate: () => {
          this.addContainerClass(target)
          this.focusInput(container)
        },
      })

      this.handleDocumentClick = (e) => {
        const node = e.target as Node
        if (!container.contains(node) && !target.contains(node)) {
          this.destroy()
        }
      }
      // add EventListener for checking if click was done outside of editor
      // only if popper has class 'context-menu'
      if (contents.classList.contains('context-menu')) {
        window.addEventListener('click', this.handleDocumentClick)
      }
    })
  }

  public destroy() {
    if (this.activePopper) {
      console.log(
        'destroying popper',
        this.activePopper,
        this.activePopper.state.elements.reference
      )
      this.removeContainerClass(
        this.activePopper.state.elements.reference as Element
      )
      this.activePopper.destroy()
      this.activePopper.state.elements.popper.remove()
      if (this.handleDocumentClick) {
        window.removeEventListener('click', this.handleDocumentClick)
      }

      delete this.activePopper
    }
  }

  public update() {
    if (this.activePopper) {
      this.activePopper.update()
    }
  }

  private focusInput(container: HTMLDivElement) {
    const element = container.querySelector('input') as HTMLDivElement | null

    if (element) {
      element.focus()
    }
  }

  private addContainerClass(referenceElement: Element) {
    const container = referenceElement.closest('.ProseMirror')

    if (container) {
      container.classList.add('popper-open')
    }
  }

  private removeContainerClass(referenceElement: Element) {
    const container = referenceElement.closest('.ProseMirror')

    if (container) {
      container.classList.remove('popper-open')
    }
  }
}
