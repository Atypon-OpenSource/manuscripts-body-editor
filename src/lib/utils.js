/*!
 * Â© 2019 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { isElementNode, } from '@manuscripts/manuscript-transform';
import { findParentNode } from 'prosemirror-utils';
export function* iterateChildren(node, recurse = false) {
    for (let i = 0; i < node.childCount; i++) {
        const child = node.child(i);
        yield child;
        if (recurse) {
            for (const grandchild of iterateChildren(child, true)) {
                yield grandchild;
            }
        }
    }
}
export const getMatchingChild = (parent, matcher) => {
    for (const node of iterateChildren(parent)) {
        if (matcher(node)) {
            return node;
        }
    }
};
export const getChildOfType = (parent, nodeType) => !!getMatchingChild(parent, node => node.type === nodeType);
export const findParentNodeWithId = findParentNode(node => 'id' in node.attrs);
export const findParentNodeWithIdValue = findParentNode(node => node.attrs.id);
export const findParentSection = findParentNode(node => node.type === node.type.schema.nodes.section);
export const findParentElement = findParentNode(node => isElementNode(node) && node.attrs.id);
