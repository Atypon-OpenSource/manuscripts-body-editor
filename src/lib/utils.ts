/*!
 * Â© 2019 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  BibliographyItemAttrs,
  BibliographyItemType,
  isElementNodeType,
  isSectionNodeType,
  ManuscriptEditorState,
  ManuscriptNode,
  ManuscriptNodeType,
  schema,
} from '@manuscripts/transform'
import {
  Fragment,
  Node as ProseMirrorNode,
  NodeType,
  ResolvedPos,
} from 'prosemirror-model'
import { EditorState, Selection } from 'prosemirror-state'
import {
  findChildrenByType,
  findParentNode,
  findParentNodeOfTypeClosestToPos,
} from 'prosemirror-utils'

import { fieldConfigMap } from '../components/references/ReferenceForm/config'
import { arrowDown } from '../icons'
import { getEditorProps } from '../plugins/editor-props'
import { isShadowDelete } from './track-changes-utils'

export function* iterateChildren(
  node: ManuscriptNode,
  recurse = false
): Iterable<ManuscriptNode> {
  for (let i = 0; i < node.childCount; i++) {
    const child = node.child(i)
    yield child

    if (recurse) {
      for (const grandchild of iterateChildren(child, true)) {
        yield grandchild
      }
    }
  }
}

export const getMatchingChild = (
  parent: ManuscriptNode,
  matcher: (node: ManuscriptNode) => boolean,
  deep = false
): ManuscriptNode | undefined => {
  for (const node of iterateChildren(parent, deep)) {
    if (matcher(node)) {
      return node
    }
  }
}

export const getMatchingDescendant = (
  parent: ManuscriptNode,
  matcher: (node: ManuscriptNode) => boolean
): ManuscriptNode | undefined => {
  for (const node of iterateChildren(parent, true)) {
    if (matcher(node)) {
      return node
    }
  }
}

export const getChildOfType = (
  parent: ManuscriptNode,
  nodeType: ManuscriptNodeType,
  deep = false
): boolean =>
  !!getMatchingChild(parent, (node) => node.type.name === nodeType.name, deep)

export const findParentNodeWithId = findParentNode((node) => 'id' in node.attrs)

export const findParentNodeWithIdValue = findParentNode((node) => node.attrs.id)

export const findParentSection = findParentNode((node) =>
  isSectionNodeType(node.type)
)

export const findParentElement = (selection: Selection, validIds?: string[]) =>
  findParentNode((node) => {
    // if validIds was passed and this element is not in it, then keep looking
    if (validIds && !validIds.includes(node.attrs.id)) {
      return false
    }
    return isElementNodeType(node.type) && node.attrs.id
  })(selection)

export const isChildOfNodeTypes = (
  doc: ManuscriptNode,
  pos: number,
  parentNodeTypes: NodeType[]
) => {
  const resolvedPos = doc.resolve(pos)
  // Iterate through the parent nodes
  for (let depth = resolvedPos.depth; depth >= 0; depth--) {
    if (parentNodeTypes.includes(resolvedPos.node(depth).type)) {
      return true
    }
  }

  return false
}

export const isSelectionInNode = (
  state: EditorState,
  targetNode: ProseMirrorNode
) => {
  const resolvedFrom = state.doc.resolve(state.selection.from)

  for (let depth = resolvedFrom.depth; depth >= 0; depth--) {
    const node = resolvedFrom.node(depth)
    if (node === targetNode) {
      return true
    }
  }
  return false
}

export const isSelectionInBody = (state: EditorState): boolean => {
  return isSelectionInNodeByType(state, 'body')
}

const isSelectionInNodeByType = (
  state: EditorState,
  nodeType: string
): boolean => {
  const { $from } = state.selection

  for (let depth = $from.depth; depth >= 0; depth--) {
    if ($from.node(depth).type.name === nodeType) {
      return true
    }
  }

  return false
}

export const createHeader = (typeName: string, text: string) => {
  const header = document.createElement('h1')
  header.classList.add(`title-${typeName}`, 'authors-info-header')
  header.textContent = text
  return header
}

export const isNotNull = <T>(a: T | null): a is T => a !== null

export const hasParent = (
  $pos: ResolvedPos,
  type: ManuscriptNodeType | ManuscriptNodeType[]
) => {
  if (Array.isArray(type)) {
    return type.some(
      (nodeType) => !!findParentNodeOfTypeClosestToPos($pos, nodeType)
    )
  }
  return !!findParentNodeOfTypeClosestToPos($pos, type)
}

// It will check if the field should be rendered based on selected item type
// and field name
export const shouldRenderField = (
  field: string,
  type: BibliographyItemType
): boolean => {
  return fieldConfigMap[type]?.has(field) ?? false
}

// It will clean unnecessary fields from the item
// id and type will be kept
export const cleanItemValues = (item: BibliographyItemAttrs) => {
  const type = item.type as BibliographyItemType
  const cleanedItem: BibliographyItemAttrs = { ...item }

  for (const key of Object.keys(item) as (keyof BibliographyItemAttrs)[]) {
    if (!shouldRenderField(key, type)) {
      switch (key) {
        case 'id':
        case 'type':
          break
        case 'author':
        case 'editor':
        case 'issued':
        case 'accessed':
        case 'event-date':
          cleanedItem[key] = undefined
          break
        default:
          cleanedItem[key] = ''
      }
    }
  }
  return cleanedItem
}

export const isBodyLocked = (state: EditorState) => {
  const props = getEditorProps(state)
  return (
    !!findChildrenByType(state.doc, schema.nodes.attachment).length &&
    props.lockBody
  )
}

// It checks if the selection is inside a body node and if the body is locked
// the body is locked if feature lockBody is set true and there is an attachment node in document
export const isEditAllowed = (state: EditorState) => {
  return !(isBodyLocked(state) && isSelectionInBody(state))
}

export const createToggleButton = (listener: () => void) => {
  const altTitlesButton = document.createElement('button')
  altTitlesButton.classList.add('toggle-button-open', 'button-reset')
  altTitlesButton.innerHTML = arrowDown
  altTitlesButton.addEventListener('click', (e) => {
    e.preventDefault()
    listener()
  })
  return altTitlesButton
}

export const getInsertPos = (
  type: ManuscriptNodeType,
  parent: ManuscriptNode,
  pos: number
) => {
  let insertPos = pos + parent.nodeSize - 1

  parent.forEach((child, offset, index) => {
    if (parent.canReplaceWith(index, index, type)) {
      insertPos = pos + offset
    }
  })

  return insertPos
}

export const filterBlockNodes = (
  fragment: Fragment,
  predicate: (node: ProseMirrorNode) => boolean
) => {
  const updatedNodes: ProseMirrorNode[] = []

  fragment.forEach((child) => {
    if (!child.isBlock) {
      updatedNodes.push(child)
      return
    }

    const newContent = child.content.size
      ? filterBlockNodes(child.content, predicate)
      : child.content
    if (predicate(child)) {
      updatedNodes.push(child.type.create(child.attrs, newContent, child.marks))
    }
  })

  return Fragment.fromArray(updatedNodes)
}
export const getLastTitleNode = (state: ManuscriptEditorState) => {
  const altTitleNode = findChildrenByType(
    state.doc,
    state.schema.nodes.alt_titles
  )[0]
  if (altTitleNode) {
    return altTitleNode
  }

  const titleNode = findChildrenByType(state.doc, state.schema.nodes.title)[0]
  return titleNode
}

/** traverse viewable nodes, that are not tracked as delete with moveNodeId */
export const visibleDescendants = (
  doc: ProseMirrorNode,
  callback: (
    node: ProseMirrorNode,
    pos: number,
    parent: ProseMirrorNode | null,
    index: number
  ) => void | boolean
) => {
  doc.descendants((node, pos, parent, index) => {
    if (isShadowDelete(node)) {
      return false
    }
    callback(node, pos, parent, index)
  })
}
