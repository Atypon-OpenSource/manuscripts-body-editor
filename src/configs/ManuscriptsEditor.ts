/*!
 * Â© 2019 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import 'prosemirror-view/style/prosemirror.css'

import { SectionCategory, UserProfile } from '@manuscripts/json-schema'
import { Capabilities } from '@manuscripts/style-guide'
import { ManuscriptNode, schema } from '@manuscripts/transform'
import { EditorState } from 'prosemirror-state'
import { EditorView } from 'prosemirror-view'
import { Location, NavigateFunction } from 'react-router-dom'
import { DefaultTheme } from 'styled-components'

import { CollabProvider } from '../classes/collabProvider'
import { clipboardParser } from '../clipboard'
import { Dispatch } from '../commands'
import { transformCopied } from '../lib/copy'
import { FileAttachment, FileManagement } from '../lib/files'
import { handleScrollToBibliographyItem } from '../lib/helpers'
import { handlePaste, transformPasted } from '../lib/paste'
import { PopperManager } from '../lib/popper'
import plugins from './editor-plugins'
import views from './editor-views'

export type CSLProps = {
  style?: string
  locale?: string
}

export interface EditorProps {
  attributes?: { [key: string]: string }
  locale: string
  theme: DefaultTheme
  getCurrentUser: () => UserProfile

  projectID: string
  doc: ManuscriptNode
  getFiles: () => FileAttachment[]
  fileManagement: FileManagement

  popper: PopperManager

  getCapabilities: () => Capabilities
  userID: string
  debug: boolean
  cslProps: CSLProps
  sectionCategories: SectionCategory[]
  collabProvider?: CollabProvider
  navigate: NavigateFunction
  location: Location
  dispatch?: Dispatch
}

export type ExternalProps = Omit<EditorProps, 'popper' | 'dispatch'>

export const createEditorState = (props: EditorProps) =>
  EditorState.create({
    doc: props.doc,
    schema,
    plugins: plugins(props),
  })

export const createEditorView = (
  props: EditorProps,
  root: HTMLElement,
  state: EditorState,
  dispatch: Dispatch
) => {
  const view = new EditorView(root, {
    state,
    editable: () => props.getCapabilities().editArticle,
    scrollMargin: {
      top: 100,
      bottom: 100,
      left: 0,
      right: 0,
    },
    dispatchTransaction: dispatch,
    nodeViews: views(props, dispatch),
    attributes: props.attributes,
    transformPasted,
    handlePaste,
    clipboardParser,
    handleScrollToSelection: handleScrollToBibliographyItem,
    transformCopied,
    handleClickOn: (view, pos, node, nodePos, event) => {
      // This to prevent changing editor selection when clicking on table cell context menu button
      if (
        event?.target &&
        (event.target as HTMLElement).classList.contains(
          'table-context-menu-button'
        )
      ) {
        return true
      }
    },
  })

  // running an init transaction allowing plugins to caught up with the document for the first time
  const tr = view.state.tr.setMeta('INIT', true)

  const nextState = view.state.apply(tr)
  view.updateState(nextState)
  return view
}
