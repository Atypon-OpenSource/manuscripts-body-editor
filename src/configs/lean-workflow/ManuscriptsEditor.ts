/*!
 * Â© 2019 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import 'prosemirror-view/style/prosemirror.css'
import '../../lib/smooth-scroll'

import { CitationProvider } from '@manuscripts/library'
import {
  Build,
  ManuscriptSchema,
  schema,
} from '@manuscripts/manuscript-transform'
import { BibliographyItem, Model } from '@manuscripts/manuscripts-json-schema'
import { Capabilities } from '@manuscripts/style-guide'
import { EditorState, Plugin } from 'prosemirror-state'
import { EditorView } from 'prosemirror-view'
import React from 'react'

import { transformPasted } from '../../lib/paste'
import { CreateView } from '../../useEditor'
import plugins from './editor-plugins-lw'
import views from './editor-views-lw'
import { ViewerProps } from './ManuscriptsViewer'

export interface EditorProps extends ViewerProps {
  plugins?: Array<Plugin<ManuscriptSchema>>
  getCitationProvider: () => CitationProvider | undefined
  saveModel: <T extends Model>(model: T | Build<T> | Partial<T>) => Promise<T>
  deleteModel: (id: string) => Promise<string>
  setLibraryItem: (item: BibliographyItem) => void
  matchLibraryItemByIdentifier: (
    item: BibliographyItem
  ) => BibliographyItem | undefined
  filterLibraryItems: (query: string) => Promise<BibliographyItem[]>
  removeLibraryItem: (id: string) => void
  setCommentTarget: (commentTarget?: string) => void
  setSelectedComment: (id?: string) => void
  retrySync: (componentIDs: string[]) => Promise<void>

  components: Record<string, React.ComponentType<any>> // eslint-disable-line @typescript-eslint/no-explicit-any
  environment?: string
  submissionId: string
  updateDesignation: (designation: string, name: string) => Promise<any> // eslint-disable-line @typescript-eslint/no-explicit-any
  uploadAttachment: (designation: string, file: File) => Promise<any> // eslint-disable-line @typescript-eslint/no-explicit-any
  capabilities?: Capabilities
  jupyterConfig: {
    url: string
    token: string
    disabled: boolean
  }
}

export default {
  createState: (props: EditorProps) => {
    return EditorState.create<ManuscriptSchema>({
      doc: props.doc,
      schema,
      plugins: plugins(props),
    })
  },

  createView: (props: EditorProps): CreateView => (el, state, dispatch) =>
    new EditorView(el, {
      state,
      editable: () => props.permissions.write,
      scrollMargin: {
        top: 100,
        bottom: 100,
        left: 0,
        right: 0,
      },
      dispatchTransaction: dispatch,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      nodeViews: views(props, dispatch) as any,
      attributes: props.attributes,
      transformPasted,
    }),
}
