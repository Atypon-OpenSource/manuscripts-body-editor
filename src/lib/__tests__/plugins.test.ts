/*!
 * Â© 2019 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { ObjectTypes, UserProfile } from '@manuscripts/json-schema'
import { getAllPermitted } from '@manuscripts/style-guide'
import { ActualManuscriptNode, schema } from '@manuscripts/transform'
import { createMemoryHistory } from 'history'
import { EditorState } from 'prosemirror-state'
import { EditorView } from 'prosemirror-view'

import plugins from '../../configs/editor-plugins'
import { EditorProps } from '../../configs/ManuscriptsEditor'
import { PopperManager } from '../popper'
import { getMatchingDescendant } from '../utils'
import jsonDoc from './__fixtures__/doc.json'
const userProfile: UserProfile = {
  _id: 'MPUserProfile:1',
  objectType: ObjectTypes.UserProfile,
  createdAt: 0,
  updatedAt: 0,
  bibliographicName: {
    _id: 'MPBibliographicName:1',
    objectType: ObjectTypes.BibliographicName,
  },
  userID: 'User|test@example.com',
}

const history = createMemoryHistory()

const buildProps = (doc: ActualManuscriptNode): EditorProps => ({
  doc,
  locale: 'en-US',
  popper: new PopperManager(),
  projectID: '',
  getCurrentUser: () => userProfile,
  history,
  setComment: () => undefined,
  setSelectedComment: () => undefined,
  setEditorSelectedSuggestion: () => undefined,
  theme: {},
  getFiles: () => [],
  fileManagement: {
    download: () => undefined, // eslint-disable-line @typescript-eslint/no-empty-function
    //@ts-ignore
    upload: () => undefined, // eslint-disable-line @typescript-eslint/no-empty-function
  },
  getCapabilities: () => getAllPermitted(),
  cslProps: {
    style: '',
    locale: '',
  },
})

describe('editor view', () => {
  const doc = schema.nodeFromJSON(jsonDoc) as ActualManuscriptNode
  beforeAll(() => {
    const root = document.createElement('div')
    root.id = 'root'
    window.document.body.appendChild(root)
  })

  test('loads without plugins', () => {
    const state = EditorState.create({
      doc,
      schema,
    })

    const view = new EditorView(null, { state })

    view.dispatch(view.state.tr.setMeta('update', true)) // trigger plugins

    const emptySection = view.state.doc.child(2).child(0)
    expect(emptySection.attrs.id).toBe('empty-section-1')
    expect(emptySection.childCount).toBe(1)

    const sectionWithEmptyParagraphs = view.state.doc.child(2).child(3)
    expect(sectionWithEmptyParagraphs.attrs.id).toBe(
      'section-with-empty-paragraphs-1'
    )
    expect(sectionWithEmptyParagraphs.childCount).toBe(3)

    const sectionWithList = view.state.doc.child(2).child(6)
    expect(sectionWithList.attrs.id).toBe('section-with-list')
    expect(sectionWithList.childCount).toBe(2)
  })

  test('loads with plugins', () => {
    const state = EditorState.create({
      doc,
      schema,
      plugins: plugins(buildProps(doc)),
    })

    const view = new EditorView(null, { state })

    view.dispatch(view.state.tr.setMeta('INIT', true)) // trigger plugins

    const sectionWithEmptyParagraphs = view.state.doc.child(2).child(3)
    expect(sectionWithEmptyParagraphs.attrs.id).toBe(
      'section-with-empty-paragraphs-1'
    )
    expect(sectionWithEmptyParagraphs.childCount).toBe(2)
    expect(sectionWithEmptyParagraphs.content.child(0).type).toBe(
      schema.nodes.section_title
    )
    expect(sectionWithEmptyParagraphs.content.child(1).type).toBe(
      schema.nodes.paragraph
    )
    expect(sectionWithEmptyParagraphs.content.child(1).textContent).toBe('')

    const sectionWithList = view.state.doc.child(2).child(6)
    expect(sectionWithList.attrs.id).toBe('section-with-list')
    expect(sectionWithList.childCount).toBe(2)
    expect(sectionWithList.content.child(0).type).toBe(
      sectionWithList.type.schema.nodes.section_title
    )

    const orderedList = sectionWithList.content.child(1)
    expect(orderedList.type).toBe(sectionWithList.type.schema.nodes.list)
    expect(orderedList.textContent).toBe('Test ')

    const inlineEquation = getMatchingDescendant(
      orderedList,
      (node) => node.type === node.type.schema.nodes.inline_equation
    )

    expect(inlineEquation).not.toBeUndefined()

    const sectionWithFigure = view.state.doc.child(2).child(7)
    expect(sectionWithFigure.attrs.id).toBe('section-with-figure')
    expect(sectionWithFigure.childCount).toBe(2)
    expect(sectionWithFigure.content.child(0).type).toBe(
      schema.nodes.section_title
    )

    const figureElement = sectionWithFigure.content.child(1)
    expect(figureElement.childCount).toBe(3)
    expect(figureElement.type).toBe(schema.nodes.figure_element)
    expect(figureElement.content.child(0).type).toBe(schema.nodes.figure)
    expect(figureElement.content.child(1).type).toBe(schema.nodes.figcaption)

    const sectionWithTable = view.state.doc.child(2).child(8)
    expect(sectionWithTable.attrs.id).toBe('section-with-table')
    expect(sectionWithTable.childCount).toBe(2)
    expect(sectionWithTable.content.child(0).type).toBe(
      sectionWithTable.type.schema.nodes.section_title
    )

    const tableElement = sectionWithTable.content.child(1)
    expect(tableElement.childCount).toBe(3)
    expect(tableElement.type).toBe(tableElement.type.schema.nodes.table_element)
    expect(tableElement.content.child(0).type).toBe(
      tableElement.type.schema.nodes.figcaption
    )
    expect(tableElement.content.child(1).type).toBe(
      tableElement.type.schema.nodes.table
    )
  })
})
