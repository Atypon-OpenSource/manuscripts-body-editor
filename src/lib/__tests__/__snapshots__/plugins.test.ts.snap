// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`editor view loads with plugins 1`] = `"<ul class=\\"manuscript-toc-list\\"><li class=\\"manuscript-toc-list-item\\" data-referenced-section=\\"MPSection:model-3\\" data-referenced-section-path-length=\\"1\\">1. An empty section</li><li class=\\"manuscript-toc-list-item\\" data-referenced-section=\\"MPSection:model-4\\" data-referenced-section-path-length=\\"1\\">2. An empty section</li><li class=\\"manuscript-toc-list-item\\" data-referenced-section=\\"MPSection:model-5\\" data-referenced-section-path-length=\\"1\\">3. An empty section</li><li class=\\"manuscript-toc-list-item\\" data-referenced-section=\\"MPSection:model-8\\" data-referenced-section-path-length=\\"1\\">4. A section containing empty paragraphs</li><li class=\\"manuscript-toc-list-item\\" data-referenced-section=\\"MPSection:model-11\\" data-referenced-section-path-length=\\"1\\">5. A section containing empty paragraphs</li><li class=\\"manuscript-toc-list-item\\" data-referenced-section=\\"MPSection:model-14\\" data-referenced-section-path-length=\\"1\\">6. A section containing empty paragraphs</li><li class=\\"manuscript-toc-list-item\\" data-referenced-section=\\"MPSection:model-16\\" data-referenced-section-path-length=\\"1\\">7. A section containing a list</li><li class=\\"manuscript-toc-list-item\\" data-referenced-section=\\"MPSection:model-19\\" data-referenced-section-path-length=\\"1\\">8. A section containing a figure</li><li class=\\"manuscript-toc-list-item\\" data-referenced-section=\\"MPSection:model-22\\" data-referenced-section-path-length=\\"1\\">9. A section containing a table</li></ul>"`;
